### PR TITLE
feat: add `prescale(sys::StateSpace)` and corresponding test.

### PR DESCRIPTION
**Pull request by tfoliva**
Adds a `prescale` function as mentioned in #462

### CODE REVIEW

File: test/test_matrix_comps.jl [hunk 1]
```diff
@@ -70,6 +70,12 @@ syst = similarity_transform(sys, Tr)
 @test sys.B ≈ Tr*syst.B
 @test sys.C*Tr ≈ syst.C
 
+nsys, T = prescale(sys)
+@test isdiagonal(nsys.A)
```
Comment (tfoliva) [inline test/test_matrix_comps.jl]: should be `isdiag`

**Review by tfoliva — COMMENTED** [review]
Fixes the `isdiag` test

**Comment by tfoliva** [comment]
I noticed that somehow the `test_matrix_comps` file does not recognize the change I made to `matrix_comps` when defining the new function `prescale`. I got the same behavior locally at my machine, It seems that the tests import the official ControlSystems package version from the Julia registry. How can I perform the tests considering the changes I made?

**Comment by baggepinnen** [comment]
You have forgotten to export the name of the new function, that's why it's not found in the tests. Where does the name come from? Is it a standard name?

**Review by tfoliva — COMMENTED** [review]
exported previously written functions

**Comment by tfoliva** [comment]
> You have forgotten to export the name of the new function, that's why it's not found in the tests. Where does the name come from? Is it a standard name?

Now I have exported the function. As for the name: I saw it was already used in MATLAB and GNU Octave's Control toolbox. Also Python-control lists a function with the same name in a sort of "to-do list" of MATLAB vs. Python-control compatibility.

On the other hand, I am also working on `canon` function to decompose a state space model in its canonical forms (diagonal, controlable and observable) and the diagonal form would eventually overcome the need for the prescale function. 

However, the `prescale` function seems to be called on lots of other functions in the aforementioned control libraries due to the fact that it returns a numerically robust representation of the system. I think here, it could have a similar role perhaps.

**Comment by baggepinnen** [comment]
Alright great! I'm not sure why the doc tests are failing, the output looks correct to me. In either case, the failures are unrelated to this PR

File: src/matrix_comps.jl [hunk 2]
```diff
@@ -583,6 +583,28 @@ function similarity_transform(sys::ST, T) where ST <: AbstractStateSpace
...
+Such that `Ã` is diagonal.
+Returns a new scaled state-space object and the associated transformation
+matrix.
+"""
+function prescale(sys::StateSpace)
+    d, S = eigen(sys.A)
+    A = Diagonal(d)
+    B = S\sys.B
+    C = sys.C*S
+    normalized_sys = iscontinuous(sys) ? ss(A, B, C, sys.D) : ss(A, B, C, sys.D, sys.Ts)
```
Comment (mfalt) [inline src/matrix_comps.jl:604]: 'StateSpace(A,B,C,sys.c,sys.timeevol)' would have been an approach that avoid the conditional here. We should probably also extend it to AbstractStateSpace, but there is a separate issue for that.

**Comment by olof3** [comment]
>Now I have exported the function. As for the name: I saw it was already used in MATLAB and GNU Octave's Control toolbox. Also Python-control lists a function with the same name in a sort of "to-do list" of MATLAB vs. Python-control compatibility.

The function `prescale` implemented in this PR does not seem to do the same thing as the Matlab function with the same name? E.g., prescale on a random system in Matlab does not give me a system with a diagonal A matrix.

I think the issue in #462 was that `similarity_transform` was broken, that would have been good to fix.

> On the other hand, I am also working on canon function to decompose a state space model in its canonical forms (diagonal, controlable and observable) and the diagonal form would eventually overcome the need for the prescale function.

Something like this could occasionally be useful though.

**Comment by tfoliva** [comment]
> The function `prescale` implemented in this PR does not seem to do the same thing as the Matlab function with the same name? E.g., prescale on a random system in Matlab does not give me a system with a diagonal A matrix.

It has been a long time since I don't use MATLAB, so I did not check if it returned a diagonal matrix. The name was taken as inspiration but it could be changed.

The idea behind the `prescale` function here was to take a badly conditioned system and transform it to diagonal canonical form which is more numerically robust. Which it does in fact do.

For the naming part, when the `canon` function is ready it could be called with a 'diagonal' option which would do the same thing as the `prescale` function and thus the latter function could be deleted.

**Comment by tfoliva** [comment]
I will look deeper into what `prescale` actually does in this other packages. Apparently as @olof3 said MATLAB uses a different scaling procedure, so I will try to find which one.

**Comment by baggepinnen** [comment]
Bump, is the current version of `prescale` something we'd like to keep for v1.0? It produces complex-coefficient systems in general, which is probably not what most people want?

**Pull request merged**